libcrux-ml-dsa/src/deserialize.rs

```diff
@@ -0,0 +1,221 @@
+    const GAMMA1: i32 = 1 << 17;
+    const GAMMA1_TIMES_2_BITMASK: i32 = (GAMMA1 << 1) - 1;
+
+    let mut re = PolynomialRingElement::ZERO;
+
+    for (i, bytes) in serialized.chunks_exact(9).enumerate() {
+        re.coefficients[4 * i + 0] = bytes[0] as i32;
+        re.coefficients[4 * i + 0] |= (bytes[1] as i32) << 8;
+        re.coefficients[4 * i + 0] |= (bytes[2] as i32) << 16;
+        re.coefficients[4 * i + 0] &= GAMMA1_TIMES_2_BITMASK;
+
+        re.coefficients[4 * i + 1] = (bytes[2] as i32) >> 2;
+        re.coefficients[4 * i + 1] |= (bytes[3] as i32) << 6;
+        re.coefficients[4 * i + 1] |= (bytes[4] as i32) << 14;
+        re.coefficients[4 * i + 1] &= GAMMA1_TIMES_2_BITMASK;
+
+        re.coefficients[4 * i + 2] = (bytes[4] as i32) >> 4;
+        re.coefficients[4 * i + 2] |= (bytes[5] as i32) << 4;
+        re.coefficients[4 * i + 2] |= (bytes[6] as i32) << 12;
+        re.coefficients[4 * i + 2] &= GAMMA1_TIMES_2_BITMASK;
+
+        re.coefficients[4 * i + 3] = (bytes[6] as i32) >> 6;
+        re.coefficients[4 * i + 3] |= (bytes[7] as i32) << 2;
+        re.coefficients[4 * i + 3] |= (bytes[8] as i32) << 10;
+        re.coefficients[4 * i + 3] &= GAMMA1_TIMES_2_BITMASK;
+
+        re.coefficients[4 * i + 0] = GAMMA1 - re.coefficients[4 * i + 0];
+        re.coefficients[4 * i + 1] = GAMMA1 - re.coefficients[4 * i + 1];
+        re.coefficients[4 * i + 2] = GAMMA1 - re.coefficients[4 * i + 2];
+        re.coefficients[4 * i + 3] = GAMMA1 - re.coefficients[4 * i + 3];
+    }
+
+    re
+}
+
+#[inline(always)]
+fn deserialize_to_mask_when_gamma1_is_2_pow_19(serialized: &[u8]) -> PolynomialRingElement {
+    const GAMMA1: i32 = 1 << 19;
+    const GAMMA1_TIMES_2_BITMASK: i32 = (GAMMA1 << 1) - 1;
+
+    let mut re = PolynomialRingElement::ZERO;
+
+    for (i, bytes) in serialized.chunks_exact(5).enumerate() {
+        re.coefficients[2 * i + 0] = bytes[0] as i32;
+        re.coefficients[2 * i + 0] |= (bytes[1] as i32) << 8;
+        re.coefficients[2 * i + 0] |= (bytes[2] as i32) << 16;
+        re.coefficients[2 * i + 0] &= GAMMA1_TIMES_2_BITMASK;
+
+        re.coefficients[2 * i + 1] = (bytes[2] as i32) >> 4;
+        re.coefficients[2 * i + 1] |= (bytes[3] as i32) << 4;
+        re.coefficients[2 * i + 1] |= (bytes[4] as i32) << 12;
+
+        re.coefficients[2 * i + 0] = GAMMA1 - re.coefficients[2 * i + 0];
+        re.coefficients[2 * i + 1] = GAMMA1 - re.coefficients[2 * i + 1];
+    }
+
+    re
+}
+
+#[inline(always)]
+pub(crate) fn deserialize_to_mask_ring_element<const GAMMA1_EXPONENT: usize>(
+    serialized: &[u8],
+) -> PolynomialRingElement {
+    match GAMMA1_EXPONENT {
+        17 => deserialize_to_mask_when_gamma1_is_2_pow_17(serialized),
+        19 => deserialize_to_mask_when_gamma1_is_2_pow_19(serialized),
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_deserialize_to_mask_when_gamma1_is_2_pow_17() {
+        let bytes = [
```

Comment:
Please add a description where this is coming from or instructions how to generate it.